### PR TITLE
feat: rename statement mutation added to synthesizer

### PR DESCRIPTION
**Pull request by hsharriman**
Added the ability for the synthesizer to randomly swap functions/constructors/predicates that have the same signatures (i.e. they take the same number and type of parameters and return the same output type, if applicable). For example, `IsSubset(B,A)` might get mutated to `Equal(B,A)` because they take two Sets as their parameters and therefore have matching signatures. As part of the ongoing synthesizer project, this broadens the scope of possible mutations that the synthesizer can choose from when generating distractors. This change adds no additional dependencies.

# Implementation strategy and design decisions

I followed the implementation strategy used for the `SwapArgs` edit functionality that was previously implemented. This presents future opportunities for refactoring, which I avoided doing too much in this PR in order to make the changes easier to follow. As a result, the current implementation plugs in with the rest of the synthesizer workflow.

# Examples with steps to reproduce them
Follow these instructions, choosing either of the indicated yarn scripts (`yarn set-example` or `yarn collinear-example`). Some of the results viewed on the generated web page will contain renamed statements (i.e. in `set-example` the added mutation would change `IsSubset(B,A)` to `Equal(B,A)`.
# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# Open questions
Where do I run `yarn docs` from? It was listed on the checklist above ^
Questions that require more discussion or to be addressed in future development:
* How should we create deterministic (or quasi-deterministic) tests for this package?
* While testing, I encountered an error where the deletion of a Bind resulted in a style compiler error being thrown. Addressing how the synthesizer will recover from user error or adjust for compiler errors that may happen downstream should be a topic for future discussion.
* Also encountered a runtime error where multiple edit mutations would sometimes attempt to mutate outdated statements. For instance: 
Screenshot from 2021-06-01 14-07-26
* Some additional refactoring to improve reusability might be a good idea as well

### CODE REVIEW

File: packages/core/src/synthesis/Synthesizer.ts [hunk 1]
```diff
@@ -381,6 +386,22 @@ export class Synthesizer {
...
+            if (newStmt.name !== stmt.name) {
+              this.cxt.replaceStmt(stmt, newStmt as ApplyPredicate);
+            }
+          } else if (stmt.tag === "Bind") {
+            let newStmt = replaceStmtName(stmt.expr as any, this.env);
+            if (newStmt.name !== (stmt.expr as ApplyConstructor).name) {
+              this.cxt.replaceStmt(stmt, {
+                ...stmt,
+                expr: newStmt,
+              } as Bind); // TODO: improve types to avoid casting
```
Comment (wodeni) [inline packages/core/src/synthesis/Synthesizer.ts:401]: Seems like `SubExpr` is too general... Let's discuss how to improve the types!

**Review by wodeni — CHANGES_REQUESTED** [review]
Great job! Overall the code is very clear, readable, and nicely documented. You definitely raised the documentation standard of the team :D. I ran some examples locally, and everything works well on my end. My comments below are mostly about code style and some minor issues that are perhaps beyond of the scope of this PR.

**Comment by wodeni** [comment]
> Where do I run `yarn docs` from? It was listed on the checklist above ^

This instruction is a little outdated, but it was specifically for `@penrose/core`. I added a top-level `yarn docs` script in #579. For now, just run it in `packages/core` and see if there's any error (unlikely tho).

File: packages/core/src/analysis/SubstanceAnalysis.ts [hunk 2]
```diff
@@ -102,6 +130,130 @@ const swap = (arr: any[], a: number, b: number) =>
...
+ */
+export const getSignature = (
+  decl: PredicateDecl | ConstructorDecl | FunctionDecl
+): Signature => {
+  let argTypes: string[] = [];
+  let outType: string | undefined;
+  if (decl.args) {
+    decl.args.forEach((arg) => {
+      if (arg.type.tag === "TypeConstructor") {
+        let a = arg.type as TypeConstructor;
```
[the remark this answers is not in the document]
Comment (hsharriman) [reply packages/core/src/analysis/SubstanceAnalysis.ts:213]: Yeah, it gives me a `property "name" does not exist on type Arg` error :(

File: packages/core/src/analysis/SubstanceAnalysis.ts [hunk 3]
```diff
@@ -102,6 +130,130 @@ const swap = (arr: any[], a: number, b: number) =>
...
+export const signatureEquals = (a: Signature, b: Signature): boolean => {
+  console.log(a, b);
+  return (
+    a.output === b.output &&
+    a.args.length === b.args.length &&
+    a.args.every((val, index) => val === b.args[index])
+  );
+};
+
+export const printStmts = (
```
[the remark this answers is not in the document]
Comment (hsharriman) [reply packages/core/src/analysis/SubstanceAnalysis.ts:246]: This is such a nifty one-liner! `stmts` is an array, so it complained that `.map()` was not a function.

File: packages/core/src/analysis/SubstanceAnalysis.ts [hunk 4]
```diff
@@ -102,6 +130,130 @@ const swap = (arr: any[], a: number, b: number) =>
...
+  stmtName: string,
+  opts: Map<string, any>
+): any[] => {
+  let matches: any[] = [];
+  // find signature of original statement in map
+  let orig = opts.get(stmtName);
+  if (orig) {
+    //generate signature for the original statement
+    var origSignature = getSignature(orig);
+    opts.map((decl) => {
```
[the remark this answers is not in the document]
Comment (hsharriman) [reply packages/core/src/analysis/SubstanceAnalysis.ts:151]: I tried a few versions of this and had issues with the iterator value complicating types that did not overlap with `ArgStmtDecl`

**Review by wodeni — APPROVED** [review]

**Pull request merged**